src/infinitempomatrix.jl

```diff
@@ -44,111 +44,212 @@ function InfiniteMPOMatrix(data::Vector{Matrix{ITensor}}, translator::Function)
+    new_ind = Index(is.space - 2; tags=new_tags)
+    mat = zeros(new_ind.space, is.space)
+    for x in 1:(new_ind.space)
+      mat[x, x + 1] = 1
+    end
+    middle = ITensor(copy(mat), new_ind, dag(is))
+  else
+    new_ind = Index(is.space[2:(end - 1)]; dir=dir(is), tags=new_tags)
+    middle = ITensors.BlockSparseTensor(
+      Float64,
+      undef,
+      Block{2}[Block(x, x + 1) for x in 1:length(new_ind.space)],
+      (new_ind, dag(is)),
+    )
+    for x in 1:length(new_ind.space)
+      dim_block = new_ind.space[x][2]
+      ITensors.blockview(middle, Block(x, x + 1)) .= diagm(0 => ones(dim_block))
+    end
+    middle = itensor(middle)
+  end
+  return top, middle, bottom
 end
 
-function find_all_links(Hms::InfiniteMPOMatrix)
-  is = inds(Hms[1][1, 1]) #site inds
-  ir = noncommonind(Hms[1][2, 1], is) #This op should have one link index pointing to the right.
-  #
-  #  Set up return array of 2-tuples
-  #
-  indexT = typeof(ir)
-  TupleT = NTuple{2,indexT}
-  inds_array = Vector{TupleT}[]
-  #
-  #  Make a dummy index
-  #
-  il0 = Index(ITensors.trivial_space(ir); dir=dir(dag(ir)), tags="Link,l=0")
-  #
-  #  Loop over sites and nonzero matrix elements which are linked into the next
-  #  and previous iMPOMatrix.  
-  #
-  for n in 1:nsites(Hms)
-    Hm = Hms[n]
-    inds_n = TupleT[]
-    lx, ly = size(Hm)
-    @assert lx == ly
-    for iy in (ly - 1):-1:1
-      ix = iy + 1
-      il = ix < lx ? commonind(Hm[ix, iy], Hms[n - 1][ix + 1, iy + 1]) : dag(il0)
-      ir = iy > 1 ? commonind(Hm[ix, iy], Hms[n + 1][ix - 1, iy - 1]) : il0
-      push!(inds_n, (il, ir))
+"""
+    convert_itensor_33matrix(tensor; leftdir=ITensors.In, kwargs...)
+
+  Converts a 4-legged tensor (coming from an (infinite) MPO) with two site indices and a left and a right leg into a 3 x 3 matrix of ITensor.
+  We assume the normal form for MPO (before full compression) where the top left and bottom right corners are identity matrices.
+  The goal is to write the tensor in the form
+         1      M_12   M_13
+         M_21   M_22   M_23
+         M_31   M_32   1
+  such that we can then easily compress it. Note that for most of our tensors, the upper triangular part will be 0.
+  Input: tensor the four leg tensors
+  Output: the 3x3 matrix of tensors
+  Optional arguments: leftdir: the direction of the left_index of the matrix to select it properly. Default to Itensors.In
+                      leftindex: instead one can directly provide the left index. Default to nothing
+                      siteindex: instead of relying on the tag, one can provide the two site indices. Default to tag filtering.
+                      left_tags: if we want to change the tags of the left indices. Default to tags(leftindex).
+                      right_tags: if we want to change the tags of the right indices. Default to tags(rightindex).
+"""
+function convert_itensor_33matrix(tensor; kwargs...)
+  @assert order(tensor) == 4
+  leftind = get(kwargs, :leftindex, nothing)
```

Comment:
We are generally moving away from the style of using `get` to access keyword arguments, instead you should define the function like:
```julia
function convert_itensor_33matrix(tensor; leftindex=nothing, ...)
```

Reply:
Ah, I did not know that. 
I was precisely forcing myself to use kwargs as I was seeing it in your code.
I will update these functions accordingly.

Do you have a document for these conventions?

Reply:
Hi Loic, There is an excellent writeup on argument handling here: https://itensor.github.io/ITensors.jl/dev/DeveloperGuide.html